parser/src/lib.rs

```diff
@@ -116,9 +116,9 @@ pub fn run_parser(
         if errors.len() > 0 {
             Err((file_library, errors))
         } else {
-            let (main_id, main_component) = main_components.pop().unwrap();
+            let (main_id, main_component, custom_gates) = main_components.pop().unwrap();
             let result_program_archive =
-                ProgramArchive::new(file_library, main_id, main_component, definitions);
+                ProgramArchive::new(file_library, main_id, main_component, definitions, custom_gates);
```

Comment:
ProgramArchive now has a new field that tells if the main component is using the custom templates pragma (which is equivalent to having declared any custom template, although it might not be used later!)